# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: netwerk/protocol/http/nsHttpConnection.cpp
# Commit: db100843b0f0
# Full Hash: db100843b0f00b89944d4dcbcb2439a48b33eaab
# Author: Valentin Gosu <valentin.gosu@gmail.com>
# Date: 2024-12-03 21:40:41
# Regressor Bug: 1929156
# File Overlap Count: 1
# Description:
#   Bug 1929156 - Check negotiated ALPN matches altSvc protocol r=necko-reviewers,kershaw
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D228217
# ==============================================================================

diff -r 16c36964a84b -r db100843b0f0 netwerk/protocol/http/nsHttpConnection.cpp
--- a/netwerk/protocol/http/nsHttpConnection.cpp	Tue Dec 03 13:50:03 2024 +0000
+++ b/netwerk/protocol/http/nsHttpConnection.cpp	Tue Dec 03 14:06:27 2024 +0000
@@ -2419,6 +2419,17 @@
   DebugOnly<nsresult> rvDebug = securityInfo->GetNegotiatedNPN(negotiatedNPN);
   MOZ_ASSERT(NS_SUCCEEDED(rvDebug));
 
+  nsAutoCString transactionNPN;
+  transactionNPN = mConnInfo->GetNPNToken();
+  LOG(("negotiatedNPN: %s - transactionNPN: %s", negotiatedNPN.get(),
+       transactionNPN.get()));
+  if (!transactionNPN.IsEmpty() && negotiatedNPN != transactionNPN) {
+    LOG(("Resetting connection due to mismatched NPN token"));
+    DontReuse();
+    mTransaction->Close(NS_ERROR_NET_RESET);
+    return;
+  }
+
   bool earlyDataAccepted = false;
   if (mTlsHandshaker->EarlyDataUsed()) {
     // Check if early data has been accepted.